crates/key-value-redis/src/lib.rs

```diff
@@ -0,0 +1,95 @@
+use anyhow::Result;
+use redis::{aio::Connection, AsyncCommands};
+use spin_core::{async_trait, key_value::Error};
+use spin_key_value::{log_error, Store, StoreManager};
+use std::sync::Arc;
+use tokio::sync::{OnceCell, RwLock};
+
+pub struct KeyValueRedis {
+    database_url: String,
+    connection: OnceCell<Arc<RwLock<Connection>>>,
+}
+
+impl KeyValueRedis {
+    pub fn new(database_url: String) -> Self {
+        Self {
+            database_url,
+            connection: OnceCell::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl StoreManager for KeyValueRedis {
+    async fn get(&self, _name: &str) -> Result<Arc<dyn Store>, Error> {
+        let connection = self
+            .connection
+            .get_or_try_init(|| async {
+                redis::Client::open(self.database_url.clone())?
+                    .get_async_connection()
+                    .await
+                    .map(RwLock::new)
+                    .map(Arc::new)
+            })
+            .await
+            .map_err(log_error)?;
+
+        Ok(Arc::new(RedisStore {
+            connection: connection.clone(),
+        }))
+    }
+}
+
+struct RedisStore {
+    connection: Arc<RwLock<Connection>>,
```

Comment:
Since we're only using the write lock is there a reason we'd prefer a RwLock over a plain Mutex?

Reply:
Ah good catch, thanks!